enumerable.rb

```diff
@@ -1,87 +1,163 @@
+# frozen_string_literal: true
+
+# rubocop: disable Metrics/ModuleLength
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+# module Enumerable
 module Enumerable
   def my_each
-    for i in self do
-        self.class == Array ? yield(i) : yield(i[0], i[1])
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i]) : yield(to_a[i][0], to_a[i][1])
+      i += 1
     end
+    self
   end
```

Comment:
- [ ]  when a block is given when self is a range behaves like #each calls the given block once for each element in self

```
  range = Range.new(1,5)
  my_each_output = ''
  block =  proc { |num| my_each_output += num.to_s }
  range.each(&block)
  p my_each_output # this is the wrong output. Test with the original each 
```

Reply:
Thank you for the test code, it helps a lot.